job/deploy_manager.go

```diff
@@ -0,0 +1,170 @@
+package job
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/odpf/salt/log"
+
+	"github.com/odpf/optimus/models"
+	"github.com/odpf/optimus/store"
+	"github.com/odpf/optimus/utils"
+)
+
+type DeployManager interface {
+	Init()
+	Deploy(ctx context.Context, projectSpec models.ProjectSpec) (models.DeploymentID, error)
+	GetStatus(ctx context.Context, deployID models.DeploymentID) (models.JobDeployment, error)
+}
+
+type deployManager struct {
+	l      log.Logger
+	config DeployManagerConfig
+
+	deployer         Deployer
+	deployerCapacity int32
+	uuidProvider     utils.UUIDProvider
+	deployRepository store.JobDeploymentRepository
+
+	// wait group to synchronise on deployers
+	wg sync.WaitGroup
+
+	// request queue, used by deployers
+	requestQ chan models.JobDeployment
+}
+
+type DeployManagerConfig struct {
+	NumWorkers    int
+	WorkerTimeout time.Duration
+	QueueCapacity int
+}
+
+func (m *deployManager) Deploy(ctx context.Context, projectSpec models.ProjectSpec) (models.DeploymentID, error) {
+	// Check deployment status for the requested Project
+	jobDeployment, err := m.deployRepository.GetByStatusAndProjectID(ctx, models.JobDeploymentStatusInQueue, projectSpec.ID)
```

Comment:
1. there are few situations where this might have a problem when optimus restarts with ongoing deployment requests, then the deployments won't happen.
2. I would suggest when optimus restarts add all the inprogress and inqueue deployment requests again to the queue, lets see if we address the issue in such a way where redoing the deployment is fine but better not to miss a deployment.

Reply:
added `Assign()` in deploy_manager that will be called as scheduled. It will:
- Assign in queue request to available deployer
- Cancel stale / timed out deployments that have been running for more than the timeout set.